### ISSUE TITLE
Errors building with Visual C++

### ISSUE THREAD

**Issue by MattHarrington** [DuffsDevice/tiny-utf8#38]
Thank you for your work on this project.  I look forward to using it.

I'm trying to use version 3.2.2 with Visual C++ inside Visual Studio Community 2019 version 16.6.0.  Here are several errors:

```
ConsoleApplication2\tinyutf8.h(3202): error C4703: potentially uninitialized local pointer variable 'app_lut_base_ptr' used
ConsoleApplication2\tinyutf8.h(3198): error C4703: potentially uninitialized local pointer variable 'old_lut_base_ptr' used
ConsoleApplication2\tinyutf8.h(3533): error C4703: potentially uninitialized local pointer variable 'str_lut_base_ptr' used
ConsoleApplication2\tinyutf8.h(3484): error C4703: potentially uninitialized local pointer variable 'old_lut_base_ptr' used
ConsoleApplication2\tinyutf8.h(3932): error C4703: potentially uninitialized local pointer variable 'repl_lut_base_ptr' used
ConsoleApplication2\tinyutf8.h(3880): error C4703: potentially uninitialized local pointer variable 'old_lut_base_ptr' used
```

Any tips for using it in Visual C++?

**Comment by DuffsDevice** [comment]
Hi Matt,

Thank you for your message!

I have created a new branch "suppress_maybe_uninitialized" that adds pragmas for MSVC to suppress the warning. May I ask you to just try it out and see if it works for you?

If that doesn't work, it seems you have supplied a compilation option, which turns warnings into errors. Therefore, I would suggest you to adjust the option that turns warnings into errors: https://stackoverflow.com/questions/2520853/warning-as-error-how-to-rid-these

Let me know, if that helps!

**Comment by MattHarrington** [comment]
I am not treating warnings as errors.  Here are the compiler flags I use, which are the default flags in MSVC 2019: `/JMC /permissive- /GS /analyze- /W3 /Zc:wchar_t /ZI /Gm- /Od /sdl /Fd"Debug\vc142.pdb" /Zc:inline /fp:precise /D "WIN32" /D "_DEBUG" /D "_CONSOLE" /D "_UNICODE" /D "UNICODE" /errorReport:prompt /WX- /Zc:forScope /RTC1 /Gd /Oy- /MDd /FC /Fa"Debug\" /EHsc /nologo /Fo"Debug\" /Fp"Debug\ConsoleApplication8.pch" /diagnostics:column`.  The flag `/WX-` indicates that warnings should not be treated as errors.

The version in the branch you mention compiles without errors.  That's great.  There are still many warnings though.

**Comment by DuffsDevice** [comment]
Oh, I see. I‘m glad though that the branch compiles finely :)
I‘ll merge it then with the master and close this ticket for now.
I didn‘t get rid of the warnings yet, as I am under GCC and only took care of the ones that occurred there. Could you nonetheless paste your full Warning output into this ticket, please?
I might be able to handle those as well, but I cannot make any guarantees, as I am very busy with work right now. Thank you,
Jakob

**Comment by MattHarrington** [comment]
Here you go:
```
tinyutf8.h(572,10): warning C4244: 'return': conversion from 'const uint64_t' to 'utf8_string::size_type', possible loss of data
tinyutf8.h(837,1): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(2156,28): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(2194,25): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(2289,28): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(2385,28): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(2473,28): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(2674,60): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(2957,38): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(3087,32): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(3212,74): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(3267,71): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(3299,74): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(3361,32): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(3543,74): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(3603,71): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(3622,70): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(3657,74): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(3744,32): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(3942,77): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(4006,71): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(4026,71): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(4069,77): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(4151,32): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(4192,94): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(4192,101): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
```

**Comment by DuffsDevice** [comment]
Thank you very much! I hope the library compiles now without warnings. Can you confirm that?

Cheers,
Jakob

**Comment by MattHarrington** [comment]
Now just two warnings when compiling with `/W3`:
```
tinyutf8.h(3757,34): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(4082,80): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
```
FYI, here are the warnings when compiling with `/W4`:
```
tinyutf8.h(714,1): warning C4127: conditional expression is constant
tinyutf8.h(714,1): message : consider using 'if constexpr' statement instead
tinyutf8.h(732,1): warning C4127: conditional expression is constant
tinyutf8.h(732,1): message : consider using 'if constexpr' statement instead
tinyutf8.h(2761,23): warning C4310: cast truncates constant value
tinyutf8.h(2762,23): warning C4310: cast truncates constant value
tinyutf8.h(2763,23): warning C4310: cast truncates constant value
tinyutf8.h(3757,34): warning C4267: 'argument': conversion from 'size_t' to 'unsigned char', possible loss of data
tinyutf8.h(4082,80): warning C4267: 'argument': conversion from 'size_t' to 'utf8_string::width_type', possible loss of data
tinyutf8.h(3162): warning C4706: assignment within conditional expression
tinyutf8.h(3442): warning C4706: assignment within conditional expression
tinyutf8.h(3829): warning C4706: assignment within conditional expression
tinyutf8.h(4122): warning C4702: unreachable code
tinyutf8.h(2952): warning C4702: unreachable code
tinyutf8.h(3049): warning C4701: potentially uninitialized local variable 'lut_width' used
tinyutf8.h(3050): warning C4701: potentially uninitialized local variable 'mb_index' used
tinyutf8.h(3220): warning C4701: potentially uninitialized local variable 'app_lut_base_ptr' used
tinyutf8.h(3216): warning C4701: potentially uninitialized local variable 'old_lut_base_ptr' used
tinyutf8.h(3352): warning C4702: unreachable code
tinyutf8.h(3551): warning C4701: potentially uninitialized local variable 'str_lut_base_ptr' used
tinyutf8.h(3710): warning C4702: unreachable code
tinyutf8.h(3502): warning C4701: potentially uninitialized local variable 'old_lut_base_ptr' used
tinyutf8.h(3950): warning C4701: potentially uninitialized local variable 'repl_lut_base_ptr' used
tinyutf8.h(3898): warning C4701: potentially uninitialized local variable 'old_lut_base_ptr' used
tinyutf8.h(1054): warning C4702: unreachable code
```

**Comment by DuffsDevice** [comment]
Could you check again after the latest commit, please? Thanks!

**Comment by MattHarrington** [comment]
No luck, but an easy fix:
```
1>C:\Source\repos\ConsoleApplication2\ConsoleApplication2\tinyutf8.h(546,65): error C2039: 'min': is not a member of 'std'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\functional(28): message : see declaration of 'std'
1>C:\Source\repos\ConsoleApplication2\ConsoleApplication2\tinyutf8.h(547,66): error C2039: 'max': is not a member of 'std'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\functional(28): message : see declaration of 'std'
1>C:\Source\repos\ConsoleApplication2\ConsoleApplication2\tinyutf8.h(553,73): error C2039: 'min': is not a member of 'std'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\functional(28): message : see declaration of 'std'
1>C:\Source\repos\ConsoleApplication2\ConsoleApplication2\tinyutf8.h(554,74): error C2039: 'max': is not a member of 'std'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\functional(28): message : see declaration of 'std'
```
Any chance of you building with MSVC yourself?  Visual Studio Community Edition is free.

**Comment by MattHarrington** [comment]
It now builds without a problem.  I also removed the `#pragma`s to disable certain MSVC warnings, and it still builds without warnings.  It's possible I updated my version of MSVC since I first opened this issue.  I'm currently on VS 16.6.1.  Maybe that's why the warnings went away.  Or, maybe you changed the code also.

I haven't investigated beyond just seeing if the header builds.  I haven't even used it yet.
